spec.bs

```diff
@@ -972,14 +752,350 @@ To <dfn>generate reporting URLs</dfn> given an {{AuctionAdConfig}} |config| and
 1. TODO: [=map/Set=] |browserSignals|["topLevelSeller"] if there was a top-level seller.
-1. TODO: Run [=evaluating script=] with [=evaluate script/script=] set to |buyerReportingScript|,
-  [=evaluate script/functionName=] set to "reportWin", and [=evaluate script/arguments=] set to
-  « |auctionSignals|, |perBuyerSignals|, |sellerSignals|, |browserSignals| ».
+1. TODO: Run [=evaluating a reporting script=] with |buyerReportingScript| "reportWin", and «
+   |auctionSignals|, |perBuyerSignals|, |sellerSignals|, |browserSignals| ».
 1. TODO: Set URLs passed to sendReportTo() into the FencedFrameConfig so they can be fetched when
   ad renders.
 
 </div>
 
+# Script runners # {#script-runners}
+
+*This introduction sub-section is non-normative*.
+
+This specification defines a new type of script execution environment called a [=script runner=]. On
+the surface, these are similar to <a href=https://html.spec.whatwg.org/C#worklets-intro>Worklets</a>
+in that they too are used for running scripts independent of the *main* execution environment with a
+flexible implementation model.
+
+However, some key differences from traditional Worklets motivate us to create a new kind of script
+execution environment. In particular, they:
+
+ * Are not scoped to a particular {{Document}}, but are rather scoped to a user agent, as they are
+   spun up by [=interest groups=] in the user agent's [=interest group set=].
+ * Consequently have a different, more flexible [=ECMAScript/agent cluster=] allocation model —
+   specifically, they need not execute in the same [=ECMAScript/agent cluster=] as any {{Document}},
+   and for security reasons implementations may be motivated to enjoy this flexibility.
+ * Do not have any other WebIDL interfaces exposed to their global besides the ones defined in this
+   specification, unlike other Worklet types.
+ * Have a restricted set of ECMAScript APIs exposed to them; for example {{Date}} is not exposed in
+   these environments.
+ * Are not [=module scripts=], and are instead evaluated as if they were [=classic scripts=].
+ * Are not [=scripts=] in the [[HTML]] sense, and as such:
+   * Their [=ECMAScript/agent=] does not have an [=event loop=]
+   * Their [=ECMAScript/realm=] does not have a [=realm/settings object=], and therefore functions
+     inside them are not run with the familiar [[WebIDL]] [=invoke|invocation=] mechanism.
+   * They do not [=perform a microtask checkpoints=].
+
+## Script evaluation ## {#script-evaluation}
+
+Concretely, a <dfn>script runner</dfn> is a JavaScript execution environment instantiated with one
+of the following global objects:
+
+ * {{InterestGroupBiddingScriptRunnerGlobalScope}}
+ * {{InterestGroupScoringScriptRunnerGlobalScope}}
+ * {{InterestGroupReportingScriptRunnerGlobalScope}}
+
+<div algorithm>
+  To <dfn>evaluate a bidding script</dfn> given a [=string=] |script| and a [=list=] of arguments
+  |arguments|:

```

Comment:
Is the order of arguments meaningful?

Reply:
I assume you mean the order of the arguments _inside_ `|arguments|`, and not the actual order of arguments _to this algorithm_ right? In that case, yes I assume so, but the callers of this algorithm are responsible for providing them in a way that the global function inside the script runner's user-provided script expects.

We should probably have an informative section that describes in developer-ese exactly what functions must be exposed to each kind of script runner and how they are called, withs some examples. I'll let @JensenPaul decide where that should be. Maybe in the "non-normative" introduction section to the `# Script runners #` section I created above? In any case, it can come after all of this infrastructure lands here.

Reply:
> I assume you mean the order of the arguments inside
Yeah, sorry that wasn't clear.

Reply:
> We should probably have an informative section that describes in developer-ese exactly what functions must be exposed to each kind of script runner and how they are called, withs some examples. 

I'm happy to propose such a section after this PR merges.  How detailed should we get in describing the arguments to methods like `generateBid()` (e.g. the Explainer has many many paragraphs of prose on the subject)?  how complex should the example `generateBid()`'s body be? i.e. just "... return { blah };" or a more elaborate example that uses some of the arguments as inputs?